heat/core/manipulations.py

```diff
@@ -911,6 +911,268 @@ def hstack(tup):
+                len(a.shape) - 1, axis
+            )
+        )
+
+    # sanitation `repeats`
+    if not isinstance(repeats, (int, list, tuple, np.ndarray, dndarray.DNDarray)):
+        raise TypeError(
+            "repeats must be a an integer, list, tuple, np.ndarray or ht.DNDarray of integers, currently: {}".format(
+                type(repeats)
+            )
+        )
+    # broadcast implied
+    if isinstance(repeats, int):
+        pass
+    else:
+        # make sure everything inside `repeats` is int
+        if isinstance(repeats, dndarray.DNDarray):
+            if repeats.dtype == types.int64:
+                pass
+            elif types.can_cast(repeats.dtype, types.int64):
+                repeats = factories.array(
+                    repeats,
+                    dtype=types.int64,
+                    is_split=repeats.split,
+                    device=repeats.device,
+                    comm=repeats.comm,
+                )
+            else:
+                raise TypeError(
+                    "Invalid dtype for ht.DNDarray 'repeats'. Has to be integer,"
+                    " but was {}".format(repeats.dtype)
+                )
+        elif isinstance(repeats, np.ndarray):
+            if not types.can_cast(repeats.dtype.type, types.int64):
+                raise TypeError(
+                    "Invalid dtype for np.ndarray 'repeats'. Has to be integer,"
+                    " but was {}".format(repeats.dtype.type)
+                )
+            repeats = factories.array(
+                repeats, dtype=types.int64, is_split=None, device=a.device, comm=a.comm
+            )
+        # invalid list/tuple
+        elif not all(isinstance(r, int) for r in repeats):
+            raise TypeError(
+                "Invalid type within repeats. All components of repeats must be integers."
+            )
+        # valid list/tuple
+        else:
+            repeats = factories.array(
+                repeats, dtype=types.int64, is_split=None, device=a.device, comm=a.comm
+            )
+
+        # check `repeats` is not empty
+        if repeats.gnumel == 0:
+            raise ValueError("Invalid input for repeats. Repeats must contain data.")
+
+        # check `repeats` is 1-dimensional
+        if len(repeats.shape) != 1:
+            raise ValueError(
+                "Invalid input for repeats. Repeats must be a 1d-object or integer, but "
+                "was {}-dimensional.".format(len(repeats.shape))
+            )
+
+    # start of algorithm
+
+    if 0 in a.gshape:
+        return a
+
+    # Broadcast (via int or 1-element DNDarray)
+    if isinstance(repeats, int) or repeats.gnumel == 1:
+        if axis is None and a.split is not None and a.split != 0:
+            print(
+                "\n!!! WARNING !!!\nIf axis is None, a has to be split along axis 0 if distributed.\nSplit axis of a "
+                "will be changed from {} to 0.".format(a.split)
+            )
+            a.resplit_(0)
```

Comment:
also, does `a` need to be resplit? could we get away with a copy and warn the user that the data is copied?

Reply:
I think a copy will be fine as well, I was just expecting the code to run faster if I do the resplit in place. Could you explain to me when to use the function directly/without the copy? Also, to which category would the warning belong? ResourceWarning? Do we have a convention on that or is it up to each developer?

Reply:
the in-place resplit will change the input matrix which can change things in future for the user. if they give a matrix to this function then try to do something later with it it can change the split of the matrix outside of the function. this is sort of bad practice, since in the other cases `a` would not be modified. you could replace this with `a = manipulations.resplit(0)` and it would probably be faster for small to large arrays (but not for very large arrays).

i think either Resource or User warnings would both be fine. its up to you